### ISSUE TITLE
Issue with aten.cat in U-Net lowering

### ISSUE THREAD

**Issue by gpetters94** [llvm/torch-mlir#945]
I'm running into an issue in U-Net, which I was using to test #917 and which I originally attributed to an issue in my code. However, looking at it more closely it appears to be an issue with `aten::cat`. I've tried to reproduce it as a test case but I wasn't able to. Here are the relevant code snippets.

Torch dialect:
```
    %148 = torch.prim.ListConstruct %147, %137 : (!torch.vtensor<[?,512,?,?],f32>, !torch.vtensor<[?,256,?,?],f32>) -> !torch.list<vtensor>
    %149 = torch.aten.cat %148, %int1 : !torch.list<vtensor>, !torch.int -> !torch.vtensor<[?,768,?,?],f32>
    %150 = torch.prim.ListConstruct  : () -> !torch.list<int>
    %false_11 = torch.constant.bool false
    %151 = torch.aten.convolution %149, %47, %none, %100, %100, %100, %false_11, %150, %int1 : !torch.vtensor<[?,768,?,?],f32>, !torch.vtensor<[256,512,3,3],f32>, !torch.none, !torch.list<int>, !torch.list<int>, !torch.list<int>, !torch.bool, !torch.list<int>, !torch.int -> !torch.vtensor<[?,256,?,?],f32>
```

Which lowers to:
```
    %147 = tensor.cast %146 : tensor<?x256x?x?xf32> to tensor<?x?x?x?xf32>
    %148 = tensor.cast %147 : tensor<?x?x?x?xf32> to tensor<?x512x?x?xf32>
    %149 = tensor.dim %148, %c0 : tensor<?x512x?x?xf32>
    %150 = tensor.dim %148, %c2 : tensor<?x512x?x?xf32>
    %151 = tensor.dim %148, %c3 : tensor<?x512x?x?xf32>
    %152 = linalg.init_tensor [%149, 768, %150, %151] : tensor<?x768x?x?xf32>
    %153 = tensor.insert_slice %148 into %152[0, 0, 0, 0] [%149, 512, %150, %151] [1, 1, 1, 1] : tensor<?x512x?x?xf32> into tensor<?x768x?x?xf32>
    %154 = tensor.dim %95, %c0 : tensor<?x256x?x?xf32>
    %155 = tensor.dim %95, %c2 : tensor<?x256x?x?xf32>
    %156 = tensor.dim %95, %c3 : tensor<?x256x?x?xf32>
    %157 = tensor.insert_slice %95 into %153[0, 512, 0, 0] [%154, 256, %155, %156] [1, 1, 1, 1] : tensor<?x256x?x?xf32> into tensor<?x768x?x?xf32>
    %158 = tensor.pad %157 low[0, 0, 1, 1] high[0, 0, 1, 1] {
    ^bb0(%arg1: index, %arg2: index, %arg3: index, %arg4: index):
      tensor.yield %cst_0 : f32
    } : tensor<?x768x?x?xf32> to tensor<?x768x?x?xf32>
    %159 = linalg.init_tensor [%149, 256, %150, %151] : tensor<?x256x?x?xf32>
    %160 = linalg.fill ins(%cst_0 : f32) outs(%159 : tensor<?x256x?x?xf32>) -> tensor<?x256x?x?xf32>
    %161 = linalg.conv_2d_nchw_fchw {dilations = dense<1> : vector<2xi64>, strides = dense<1> : vector<2xi64>} ins(%158, %cst_55 : tensor<?x768x?x?xf32>, tensor<256x512x3x3xf32>) outs(%160 : tensor<?x256x?x?xf32>) -> tensor<?x256x?x?xf32>
```

When running `torch-mlir-opt -pass-pipeline='func.func(refback-generalize-tensor-pad),func.func(scf-bufferize),func.func(tm-tensor-bufferize),func.func(linalg-bufferize),func-bufferize,arith-bufferize,func.func(tensor-bufferize),func.func(finalizing-bufferize),refback-munge-calling-conventions,refback-insert-rng-globals,func.func(tm-tensor-to-loops),func.func(refback-munge-memref-copy),func.func(convert-linalg-to-loops),func.func(lower-affine),convert-scf-to-cf,func.func(refback-expand-ops-for-llvm),func.func(arith-expand),func.func(convert-math-to-llvm),convert-linalg-to-llvm,convert-memref-to-llvm,func.func(convert-arith-to-llvm),convert-func-to-llvm,convert-cf-to-llvm,reconcile-unrealized-casts'` on the resulting MLIR, I get the following error:

```
linalg.mlir:473:12: error: 'memref.copy' op requires the same shape for all operands
    %153 = tensor.insert_slice %148 into %152[0, 0, 0, 0] [%149, 512, %150, %151] [1, 1, 1, 1] : tensor<?x512x?x?xf32> into tensor<?x768x?x?xf32>
           ^
linalg.mlir:473:12: note: see current operation: "memref.copy"(%312, %316) : (memref<?x256x?x?xf32>, memref<?x512x?x?xf32, affine_map<(d0, d1, d2, d3)[s0, s1, s2, s3] -> (d0 * s1 + s0 + d1 * s2 + d2 * s3 + d3)>>) -> ()
```

The full torch dialect input is [here](https://gist.github.com/vivekkhandelwal1/bd28c87266af79a736188dbf91c4a572)

**Comment by silvasean** [comment]
> I've tried to reproduce it as a test case but I wasn't able to.

What is the original issue? Is this memref.copy issue related?

**Comment by gpetters94** [comment]
aten.cat is implemented as multiple `insert_slice` ops, and as the error message says the memref issue is arising from the second `insert_slice` in the cat op.

**Comment by silvasean** [comment]
Can you reduce the test case to a single pass and a small (<10 ops) function?

**Comment by gpetters94** [comment]
This was an issue with a different op (transposed convolution) that I have fixed in #917.
